### PR TITLE
Optimize CI workflow

### PR DESCRIPTION
**Pull request by Hypnosphi**
Issue: #2179

### CODE REVIEW

**Comment by Hypnosphi** [comment]
I made all the other jobs depend on `build`. This makes overall workflow time for a particular commit a bit longer, but significantly diminishes total time spent on agents, which means that builds will spend less time waiting in queue.

before: https://circleci.com/workflow-run/c7e6c92c-90f4-4746-8b4a-d70c21cf4850
after: https://circleci.com/workflow-run/5317bb4f-c4e9-4163-917a-ec4885f97043

**Comment by Hypnosphi** [comment]
A further optimisation would be to move the docs and react native bootstrapping from `build` to the corresponding jobs, so that `example-kitchen-sinks` don't have to wait for it. Will try it now

**Comment by Hypnosphi** [comment]
After moving `docs` bootstrapping to its own job:
https://circleci.com/workflow-run/048fc41b-0150-4a32-9de1-5988e2112a8d

**Comment by Hypnosphi** [comment]
After moving `react-native` bootstrapping to its own job:
https://circleci.com/workflow-run/ac85393a-553f-4fbd-9ae5-0a7738e1dcd2

The next step could be moving RN unit-tests to `example-react-native` (and probably renaming the latter to just `react-native`)

**Comment by Hypnosphi** [comment]
And voila — after moving react-native tests to the corresponding job both overall workflow time and total agent time is significantly less:
https://circleci.com/workflow-run/2ca1fd0a-5edd-443d-b6d5-f3c22d669319

File: .circleci/config.yml [hunk 1]
```diff
@@ -13,43 +13,52 @@ jobs:
...
           paths:
             - node_modules
-            - docs/node_modules
-            - examples/react-native-vanilla/node_modules
-            - examples/crna-kitchen-sink/node_modules
+            - example/cra-kitchen-sink/node_modules
+            - example/vue-kitchen-sink/node_modules
+      - save_cache:
+          name: "Cache core dist"
+          key: core-dist-{{ .Revision }}
```
Comment (tmeasday) [inline .circleci/config.yml]: What does this bit do?

Comment (tmeasday) [reply .circleci/config.yml]: Does bootstrap build the tarballs now?

File: .circleci/config.yml [hunk 2]
```diff
@@ -13,43 +13,52 @@ jobs:
...
+          paths:
+            - addons
+            - app
+            - lib
```
Comment (Hypnosphi) [inline .circleci/config.yml]: This caches `app`, `addons`, and `lib` directories using current commit hash as a part of the cache key.

`bootstrap --core` transpiles all the packages in those directories (`src` to `dist`).

Ideally, it should be
```
paths:
  - addons/*/dist
  - app/*/dist
  - lib/*/dist
```

But unfortunately, wildcards are not supported in cached directories paths

**Review by danielduan — APPROVED** [review]
This is soo awesome, thanks!

**Pull request merged**